File: SDK Examples/Examples/C#_Console/Counter_BufferedPWMOutput/BufferedPWMOutput.cs
/*******************************************************************************
Copyright (c) 1983-2016 Advantech Co., Ltd.
********************************************************************************
THIS IS AN UNPUBLISHED WORK CONTAINING CONFIDENTIAL AND PROPRIETARY INFORMATION
WHICH IS THE PROPERTY OF ADVANTECH CORP., ANY DISCLOSURE, USE, OR REPRODUCTION,
WITHOUT WRITTEN AUTHORIZATION FROM ADVANTECH CORP., IS STRICTLY PROHIBITED. 
================================================================================
REVISION HISTORY
--------------------------------------------------------------------------------
$Log: $
 --------------------------------------------------------------------------------
 $NoKeywords:  $
*/
/******************************************************************************
*
* Windows  Example:
*    BufferedPWMOutput.cs
*
* Example Category:
*    Counter
*
* Description:
*    This example demonstrates how to use Buffered PWM Output function.
*
* Instructions for Running:
*    1  Set the 'deviceDescription' for opening the device. 
*	  2  Set the 'profilePath' to save the profile path of being initialized device. 
*    3  Set the 'chEnable' to enabled channel of the counter to operate
*    4  set the 'pulseWidth' to decide the period of pulse for selected channel.
*    
* I/O Connections Overview:
*    Please refer to your hardware reference manual.
*
******************************************************************************/
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using System.Reflection;
using System.Runtime.InteropServices;
using Automation.BDaq;

namespace Counter_BufferedPWMOutput
{
   class BufferedPWMOutput
   {
      static void Main(string[] args)
      {
         //-----------------------------------------------------------------------------------
         // Configure the following parameters before running the demo
         //-----------------------------------------------------------------------------------
         //The default device of project is demo device, users can choose other devices according to their needs. 
         string deviceDescription = "PCIE-1813,BID#15";
         string profilePath = "../../../profile/PCIE-1813.xml";
         int intervalCount = 2048;
         int samples = 4096;
         int userBufLength = 4096;
         PulseWidth []dataBuf = new PulseWidth[userBufLength];
         ErrorCode errorCode = ErrorCode.Success;

         // Step 1: Create a 'BufferedPwModulatorCtrl' for Buffered PWM Output function.
         BufferedPwModulatorCtrl bfdPwModulatorCtrl = new BufferedPwModulatorCtrl();

         // Step 2: Set the notification event Handler by which we can known the state of operation effectively. 
         bfdPwModulatorCtrl.DataTransmitted += new EventHandler<BfdCntrEventArgs>(bfdPwModulatorCtrl_DataTransmitted);

          try
         {
            // Step 3: Select a device by device number or device description and specify the access mode.
            // in this example we use ModeWrite mode so that we can fully control the device, including configuring, sampling, etc.
            bfdPwModulatorCtrl.SelectedDevice = new DeviceInformation(deviceDescription);
				errorCode = bfdPwModulatorCtrl.LoadProfile(profilePath);//Loads a profile to initialize the device.
            if (BioFailed(errorCode))
            {
               throw new Exception();
            }

            // Step 4: Set necessary parameters.
            int channelCountMax = bfdPwModulatorCtrl.Features.ChannelCountMax;
            byte[] chEnable = new byte[] { 1, 0, 0, 0 };
            bfdPwModulatorCtrl.ChannelMap = chEnable;

            bfdPwModulatorCtrl.SampleClkSrc = SignalDrop.SignalNone;

            bfdPwModulatorCtrl.IntervalCount = intervalCount;
            bfdPwModulatorCtrl.Samples = samples;
            bfdPwModulatorCtrl.Streaming = true;
            
             // Step 5: Set Trigger parameters
            Trigger trigger = bfdPwModulatorCtrl.Trigger;
            if (trigger != null)
            {
               trigger.Source = bfdPwModulatorCtrl.Features.TriggerSources[0];// set trigger source to 'SignalNone'.
            }

            for (int i = 0; i < userBufLength; i++)
            {
               dataBuf[i].HiPeriod = 0.8;
               dataBuf[i].LoPeriod = 0.2;
            }
            for (int i = 0; i < channelCountMax; i++ )
            {
               if (chEnable[i] == 1){
                  bfdPwModulatorCtrl.SetData(i, userBufLength, dataBuf);
               }
            }

            // Step 6: start BufferedPWMOutput
            Console.WriteLine(" BufferedPWMOutput is in progress...\n Test signal to the Out pin !");
            Console.WriteLine(" Any key to quit !\n");
            
            bfdPwModulatorCtrl.Prepare();

            bfdPwModulatorCtrl.Enabled = true;

		      while(!Console.KeyAvailable)
		      {
			      Thread.Sleep(1000);         
		      }

            // Step 7: Stop BufferedPWMOutput
            bfdPwModulatorCtrl.Enabled = false;
         }
         catch(Exception e)
         {
            // Something is wrong
            string errStr = BioFailed(errorCode) ? " Some error occurred. And the last error code is " + errorCode.ToString()
                                         : e.Message;
            Console.WriteLine(errStr);
         }
         finally
         {
            //Step 6: Close device and release any allocated resource.
            bfdPwModulatorCtrl.Dispose();
            Console.ReadKey(false);
         }
      }

      static bool BioFailed(ErrorCode err)
      {
         return err < ErrorCode.Success && err >= ErrorCode.ErrorHandleNotValid;
      }

      static void bfdPwModulatorCtrl_DataTransmitted(object sender, BfdCntrEventArgs e)
      {
         Console.WriteLine(" BufferedPWMOut Data Transmitted: channel:{0}, count:{1}", e.Channel, e.Count);
      }
   }
}

File: SDK Examples/Examples/ANSI_C/Counter_BufferedPWMOutput/BufferedPWMOutput.c
/*******************************************************************************
Copyright (c) 1983-2017 Advantech Co., Ltd.
********************************************************************************
THIS IS AN UNPUBLISHED WORK CONTAINING CONFIDENTIAL AND PROPRIETARY INFORMATION
WHICH IS THE PROPERTY OF ADVANTECH CORP., ANY DISCLOSURE, USE, OR REPRODUCTION,
WITHOUT WRITTEN AUTHORIZATION FROM ADVANTECH CORP., IS STRICTLY PROHIBITED. 
================================================================================
REVISION HISTORY
--------------------------------------------------------------------------------
$Log: $
--------------------------------------------------------------------------------
$NoKeywords:  $
*/
/******************************************************************************
*
* Windows Example:
*     BufferedPWMOutput.c
*
* Example Category:
*    Counter
*
* Description:
*    This example demonstrates how to use Buffered PWM Output function.
*
* Instructions for Running:
*    1  Set the 'deviceDescription' for opening the device. 
*	  2  Set the 'profilePath' to save the profile path of being initialized device. 
*
* I/O Connections Overview:
*    Please refer to your hardware reference manual.
*
******************************************************************************/
#include <stdlib.h>
#include <stdio.h>
#include "../inc/compatibility.h"
#include "../../../inc/bdaqctrl.h"
//-----------------------------------------------------------------------------------
// Configure the following parameters before running the demo
//-----------------------------------------------------------------------------------
#define     deviceDescription L"PCIE-1813,BID#15"
const wchar_t* profilePath = L"../../profile/PCIE-1813.xml";

#define  IntervalCount  2048
#define  Samples        4096     
#define  UserBufLength  Samples     
void waitAnyKey()
{
   do{SLEEP(1);} while(!kbhit());
} 

// This function is used to deal with 'DataTransmitted' Event.
void BDAQCALL OnDataTransmittedEvent(void * sender, BfdCntrEventArgs * args, void *userParam)
{
   printf("BufferedPWMOut Data Transmitted: channel = %d, count = %d, offset = %d\n", args->Channel, args->Count, args->Offset);
}
// This function is used to deal with 'TransStopped' Event.
void BDAQCALL OnTransStoppedEvent(void * sender, BfdCntrEventArgs * args, void *userParam)
{
   printf("BufferedPWMOutput Transmit Stopped: channel = %d, count = %d, offset = %d\n", args->Channel, args->Count, args->Offset);
}
// This function is used to deal with 'Stopped' Event.
void BDAQCALL OnStoppedEvent(void * sender, BfdCntrEventArgs * args, void *userParam)
{
   printf("BufferedPWMOutput stopped: channel = %d, count = %d, offset = %d\n", args->Channel, args->Count, args->Offset);
}

int main(int argc, char* argv[])
{
   ErrorCode ret = Success;
   Trigger* trigger = NULL;
   int32 chCount     = 4;
   int8  chEnabled[] = {1, 0, 0, 0};
   int i = 0;
   PulseWidth userData[UserBufLength] = {0};

   // Step 1: Create a 'BufferedPwModulatorCtrl' for Buffered PWM Output function.
   BufferedPwModulatorCtrl* bfdPwModulatorCtrl = BufferedPwModulatorCtrl_Create();

   // Step 2: Set the notification event Handler by which we can known the state of operation effectively.
   BufferedPwModulatorCtrl_addDataTransmittedHandler(bfdPwModulatorCtrl, OnDataTransmittedEvent, NULL);
   BufferedPwModulatorCtrl_addTransitStoppedHandler(bfdPwModulatorCtrl, OnTransStoppedEvent, NULL);
   BufferedPwModulatorCtrl_addStoppedHandler(bfdPwModulatorCtrl, OnStoppedEvent, NULL);

   do
   {
      // Step 3: Select a device by device number or device description and specify the access mode.
      // in this example we use ModeWrite mode so that we can fully control the device, including configuring, sampling, etc.
      DeviceInformation devInfo;
      devInfo.DeviceNumber = -1;
      devInfo.DeviceMode   = ModeWrite;
      devInfo.ModuleIndex  = 0;
      wcscpy(devInfo.Description, deviceDescription);
      ret = BufferedPwModulatorCtrl_setSelectedDevice(bfdPwModulatorCtrl, &devInfo);
      CHK_RESULT(ret);
      ret = BufferedPwModulatorCtrl_LoadProfile(bfdPwModulatorCtrl, profilePath);//Loads a profile to initialize the device.
      CHK_RESULT(ret);

      // Step 4: Set necessary parameters.
      ret = BufferedPwModulatorCtrl_setChannelMap(bfdPwModulatorCtrl, chCount, chEnabled);
	   CHK_RESULT(ret);

      ret = BufferedPwModulatorCtrl_setSampleClkSrc(bfdPwModulatorCtrl, SignalNone);
      CHK_RESULT(ret);

      ret = BufferedPwModulatorCtrl_setIntervalCount(bfdPwModulatorCtrl, IntervalCount);
      CHK_RESULT(ret);
      ret = BufferedPwModulatorCtrl_setSamples(bfdPwModulatorCtrl, Samples);
      CHK_RESULT(ret);
      ret = BufferedPwModulatorCtrl_setStreaming(bfdPwModulatorCtrl, TRUE);
      CHK_RESULT(ret);

      trigger = BufferedPwModulatorCtrl_getTrigger(bfdPwModulatorCtrl); //Setting trigger
      if(trigger) {
         ret = Trigger_setSource(trigger, SignalNone); // The "SignalNone" means doing without trigger.
         CHK_RESULT(ret);
      }

      for(i = 0; i < UserBufLength; i++) {
         userData[i].HiPeriod = 0.5;
         userData[i].LoPeriod = 0.5;
      }

      for(i = 0; i < chCount; i++) { 
         if(chEnabled[i]) {
            ret = BufferedPwModulatorCtrl_SetData(bfdPwModulatorCtrl, i, UserBufLength, userData);
            CHK_RESULT(ret);
         }
      }

      // Step 5: Start Buffered PWM Out Counter 
      ret = BufferedPwModulatorCtrl_Prepare(bfdPwModulatorCtrl);
      CHK_RESULT(ret);

	   ret= BufferedPwModulatorCtrl_setEnabled(bfdPwModulatorCtrl, TRUE);
	   CHK_RESULT(ret);

	   printf("PWMOutput is in progress... any key to quit !\n");
	   while ( !kbhit()) {
         SLEEP(1);
      }

	   // step 6: Stop the operation if it is running.
	   ret= BufferedPwModulatorCtrl_setEnabled(bfdPwModulatorCtrl, FALSE);
	   CHK_RESULT(ret);
   }while(FALSE);

   // Step 7: Close device, release any allocated resource.
	BufferedPwModulatorCtrl_Dispose(bfdPwModulatorCtrl);

	// If something wrong in this execution, print the error code on screen for tracking.
   if(BioFailed(ret))
   {
      printf("\nSome error occurred. And the last error code is 0x%X.\n", ret);
      waitAnyKey();
   }  
   return 0;
}
